### ISSUE TITLE
Upgrade to Java8

### ISSUE THREAD

**Issue by mattrjacobs** [Netflix/Hystrix#702]

**Comment by alexzorin** [comment]
Hi, 

Does this mean you are not supporting Java 6 as a target runtime? `README.md` currently states:
>You need Java 6 or later.

But, even with 1.4.0 release, I cannot use it with Java 6, due to an error:

```com/netflix/hystrix/HystrixCommand : Unsupported major.minor version 51.0```

Because the release jar on maven is built with Java 7:

```
$ javap -verbose Hystrix.class | grep -E "major|minor"
  minor version: 0
  major version: 51
```

Thanks

**Comment by mattrjacobs** [comment]
Yeah, my mistake in not updating the README (it's updated now).  Java7 upgrade was included in 1.4.0 (#373), and I'm currently targeting Java8 for inclusion in 1.5.0 (https://github.com/Netflix/Hystrix/wiki/Migration-Guide#proposed-changes-from-14x-to-15x)

**Comment by mattrjacobs** [comment]
Also, thanks for the report.  Good catch!

**Comment by mattrjacobs** [comment]
Note that we are now back on Java 6 as of Hystrix 1.4.3.  So this is delayed indefinitely.  

One way that this could go is that I could discover a few breaking changes I'd like to make, and then make them in a Hystrix 2.0 branch, along with all removal of deprecated features and a jump to Java 8.  

I don't have a concrete timeline here - just an idea that bundling breaking changes together in a 2.0 is likely better than doing it piecemeal through the 1.x series.

**Comment by spencergibb** [comment]
Is the goal of this Java 8 *language* features or Java 8 *APIs*?  Spring supports Java 8 API's for clients that use it, while maintaining backwards compatibility to Java 6.

**Comment by mattrjacobs** [comment]
It's a good question.  I haven't explored what Java 8 would add to the implementation of hystrix-core yet, but my hunch is that the functional Rx flow is very well-represented using lambdas.  I also think default methods might make some of the super-verbose constructor options more straightforward.

Do you have a pointer for any documentation around their approach? - I'd be interested to learn more.

**Comment by spencergibb** [comment]
I know they compile with java 8 targeting java 6.  Maybe @jhoeller, @philwebb or someone else from the core spring team could comment.

**Comment by spencergibb** [comment]
From: [Spring Framework build.gradle](https://github.com/spring-projects/spring-framework/blob/master/build.gradle#L143-L169)

```
	task sniff {
		group = "Verification"
		description = "Checks the Java API signatures"

		dependsOn compileJava
		dependsOn copyJavaApiSignature

		inputs.dir sourceSets.main.output.classesDir
		inputs.dir copyJavaApiSignature.to
		outputs.upToDateWhen { true }
		
		doLast {
			ant.taskdef(
				name: 'animalSniffer',
				classname: 'org.codehaus.mojo.animal_sniffer.ant.CheckSignatureTask',
				classpath: configurations.sniffer.asPath
			)

			ant.animalSniffer(
					signature: "$buildDir/javaApiSignature/javaApi.signature",
					classpath: sourceSets.main.compileClasspath.asPath) {
				path(path: sourceSets.main.output.classesDir)
				annotation(className: "org.springframework.lang.UsesJava7")
				annotation(className: "org.springframework.lang.UsesJava8")
				annotation(className: "org.springframework.lang.UsesSunHttpServer")
			}
		}
	}
```

Looks like they use animal sniffer and mark code that uses Java 8, etc. and enforce it.

**Comment by spencergibb** [comment]
They also run their tests against each jdk version.

**Comment by spencergibb** [comment]
Many thanks to @rwinch for the details.

**Comment by jhoeller** [comment]
Coincidentally, we're in the process of writing a blog post on this topic, scheduled for next week...
 
The gist is that we're conditionally activating Java 8 features in quite a few places, reacting to the use of Java 8 features in a user's field/method/constructor declarations (e.g. the JSR-310 date-time types, Optional, Stream, default methods). This is quite straightforward, but then that judgement comes from somebody who even bridged JDK 1.4 / 5 / 6 the same way back in the Spring 2.5 line ;-)

Lambdas are very naturally compatible with Spring's API style in general, without us having to do much about it, just through the common compiler convention in use there. In specific places, we started offering lambda-friendly single-method callbacks as an alternative to a few multi-method interfaces, but most callback interfaces in Spring were single-method interfaces to begin with.

In any case, the goal simply is: From an application developer's perspective, Spring 4.x should feel like a Java 8 based framework when working with it on Java 8... with the very same jars, without any special setup, just through auto-adapting at runtime. The only limitation that I'm aware of at this point is that we're not declaring default methods in any of the framework's interfaces but rather still provide pre-implemented abstract base classes for some of our SPIs as an implementation convenience.

Now, would I recommend this approach for other frameworks? In principle, yes, in particular when oftentimes used on existing application server platforms where people don't independently choose their JDK vendor and version. However, the stronger the benefit of Java 8 language and API features is for a framework's usage model, the more I would lean towards fully going Java 8+, at least as of 2016.

**Comment by mattrjacobs** [comment]
That's very helpful feedback, @jhoeller , thanks for sharing it.  I look forward to reading the blog post.

**Comment by rwinch** [comment]
For those on this thread interested, the blog is now posted at https://spring.io/blog/2015/04/01/ongoing-support-for-java-7-and-even-java-6

cc @mattrjacobs

**Comment by vijayeswari** [comment]
Hello Everyone,

I am new to explore Hystrix and currently working on a PoC on Hystrix.
I cloned the Hystrix source, pointing to the master, I believe its the latest version 1.4.21

I just followed README and installed Java 6 and later found out the compatibility issues with JRE versions, major and minor, as already discussed in the thread.

Later tried to install JDK 7 for Hystrix 1.4.21, but when I do a build, I get the following error:
* What went wrong:
Could not resolve all dependencies for configuration ':hystrix-rx-netty-metrics-stream:compile'.
> Could not resolve io.reactivex:rxnetty:0.4.12.
  Required by:
      com.netflix.hystrix:hystrix-rx-netty-metrics-stream:1.5.0-SNAPSHOT
   > Could not resolve io.reactivex:rxnetty:0.4.12.
      > Could not get resource 'https://jcenter.bintray.com/io/reactivex/rxnetty/0.4.12/rxnetty-0.4.12.pom'.
         > Could not GET 'https://jcenter.bintray.com/io/reactivex/rxnetty/0.4.12/rxnetty-0.4.12.pom'.
            > peer not authenticated
> Could not resolve org.codehaus.jackson:jackson-core-asl:1.9.2.
  Required by:
      com.netflix.hystrix:hystrix-rx-netty-metrics-stream:1.5.0-SNAPSHOT
   > Could not resolve org.codehaus.jackson:jackson-core-asl:1.9.2.
      > Could not get resource 'https://jcenter.bintray.com/org/codehaus/jackson/jackson-core-asl/1.9.2/jackson-core-asl-1.9.2.pom'.
         > Could not GET 'https://jcenter.bintray.com/org/codehaus/jackson/jackson-core-asl/1.9.2/jackson-core-asl-1.9.2.pom'.
            > peer not authenticated
> Could not resolve org.codehaus.jackson:jackson-mapper-asl:1.9.2.
  Required by:
      com.netflix.hystrix:hystrix-rx-netty-metrics-stream:1.5.0-SNAPSHOT
   > Could not resolve org.codehaus.jackson:jackson-mapper-asl:1.9.2.
      > Could not get resource 'https://jcenter.bintray.com/org/codehaus/jackson/jackson-mapper-asl/1.9.2/jackson-mapper-asl-1.9.2.pom'.
         > Could not GET 'https://jcenter.bintray.com/org/codehaus/jackson/jackson-mapper-asl/1.9.2/jackson-mapper-asl-1.9.2.pom'.
            > peer not authenticated

I tried with JDK 8, it gave me build errors in hystrix-javanica. I kind of feel JDK 8 would still have compatibility issues. 

So i am only left with JDK 7 and not sure how to resolve the build errors. Any help would be appreciated. Thanks.

**Comment by mattrjacobs** [comment]
@vijayeswari Thanks for the report.

The Nebula plugin does not support Java6, as mentioned above.
Java 7 worked for me, and I also retried with a clean Gradle cache.  That worked as well.  Can you re-try?
Java 8 did not work, and that was a news to me.  I opened #988 for @dmgcodevil to look at.

**Comment by vijayeswari** [comment]
@mattrjacobs  Thanks very much for the quick reply. 
I will try with clean Gradle cache and JDK 7 and update you on that.
[JDK8_Hystrix_MasterBranch_1.4.21_buildError.txt](https://github.com/Netflix/Hystrix/files/40378/JDK8_Hystrix_MasterBranch_1.4.21_buildError.txt)

**Comment by vijayeswari** [comment]
@mattrjacobs  Yes if worked with JDK 7 after clearing the Gradle cache!! Thanks so much!

**Comment by mattrjacobs** [comment]
This update is now merged into the 2.0.x branch
